amm/contracts/stable_pool/lib.rs

```diff
@@ -0,0 +1,1033 @@
+        fn get_swap_amount_out(
+            &mut self,
+            token_in: AccountId,
+            token_out: AccountId,
+            token_in_amount: u128,
+        ) -> Result<(u128, u128), StablePoolError> {
+            self.update_rates();
+            let (token_in_id, token_out_id) = self.check_tokens(token_in, token_out)?;
+            let rates = self.get_scaled_rates()?;
+            Ok(math::rated_swap_to(
+                &rates,
+                token_in_id as usize,
+                token_in_amount,
+                token_out_id as usize,
+                &self.reserves(),
+                &self.pool.fees,
+                self.amp_coef(),
+            )?)
+        }
+
+        #[ink(message)]
+        fn get_swap_amount_in(
+            &mut self,
+            token_in: AccountId,
+            token_out: AccountId,
+            token_out_amount: u128,
+        ) -> Result<(u128, u128), StablePoolError> {
+            self.update_rates();
+            let (token_in_id, token_out_id) = self.check_tokens(token_in, token_out)?;
+            let rates = self.get_scaled_rates()?;
+            Ok(math::rated_swap_from(
+                &rates,
+                token_in_id as usize,
+                token_out_amount,
+                token_out_id as usize,
+                &self.reserves(),
+                &self.pool.fees,
+                self.amp_coef(),
+            )?)
+        }
+
+        #[ink(message)]
+        fn get_mint_liquidity_for_amounts(
+            &mut self,
+            amounts: Vec<u128>,
+        ) -> Result<(u128, u128), StablePoolError> {
+            self.update_rates();
+            if amounts.len() != self.pool.tokens.len() {
+                return Err(StablePoolError::IncorrectAmountsCount);
+            }
+            let rates = self.get_scaled_rates()?;
+
+            Ok(math::rated_compute_lp_amount_for_deposit(
+                &rates,
+                &amounts,
+                &self.reserves(),
+                self.psp22.total_supply(),
+                Some(&self.pool.fees),
+                self.amp_coef(),
+            )?)
+        }
+
+        #[ink(message)]
+        fn get_amounts_for_liquidity_mint(
+            &mut self,
+            liquidity: u128,
+        ) -> Result<Vec<u128>, StablePoolError> {
+            match math::compute_deposit_amounts_for_lp(
+                liquidity,
+                &self.reserves(),
+                self.psp22.total_supply(),
+            ) {
+                Ok((amounts, _)) => Ok(amounts),
+                Err(err) => Err(StablePoolError::MathError(err)),
+            }
+        }
+
+        #[ink(message)]
+        fn get_burn_liquidity_for_amounts(
+            &mut self,
+            amounts: Vec<u128>,
+        ) -> Result<(u128, u128), StablePoolError> {
+            self.update_rates();
+            if amounts.len() != self.pool.tokens.len() {
+                return Err(StablePoolError::IncorrectAmountsCount);
+            }
```

Comment:
Doesn't make sense to update rates if this returns an error - it will revert anyway. We should ensure the lengths are equal at the very beginning.

Reply:
[Various small things](https://github.com/Cardinal-Cryptography/common-amm-stable-swap/pull/5/commits/77337f17f3b8eeb2accacb62c0c60d728ecb9814)